Councillor contact form

alexander-griffen opened the pull request:
Fixes #9 .

This imports the climate representative data from [here](https://github.com/mysociety/comeval-climate-data). Currently, this data was manually exported into csvs, and then dumped into `data/procurement_data` - I'm not sure if there's a better way of doing this? (If a better way is found, the README should be updated to reflect this).

This data is then used to identify relevant Councillors/Officers for the user based on their postcode.

The user is then directed to a form to be able to email them.

This also will need more/better copywriting.

alexander-griffen committed:
Add climate representative models

alexander-griffen committed:
Add command to download climate representative data.

alexander-griffen committed:
Tidied up forms.

This involved the following:
- Removing the now unused form used for choosing the source in the
email alerts page.
- Subclassing forms with postcodes in them to use the 'clean_pc' method
across forms.
- Updating 'filters.py' to reflect the subclassing of the 'clean_pc'
method.

alexander-griffen committed:
Create form for contacting representatives

alexander-griffen committed:
Representative contact journey, including searching for relevent
representatives based on postcode.
Currently, the form for the contacting itself takes you to the success
page without actually doing anything, as it doesn't send an email, it
just validates the form.

alexander-griffen committed:
Updated README to reflect new data import.

alexander-griffen committed:
Alter import command so that it accepts alternate data files (for
testing)

alexander-griffen committed:
Test climate representative import

alexander-griffen committed:
Councillor contact tests added

zarino committed:
Link to council contact form, from contract detail page

alexander-griffen committed:
Labels changed to be produced by bootstrap, human-friendly labels
created.

zarino committed:
Couple of styling tweaks for contact_representative form

zarino reviewed (commented):
Looks like just a couple of things left to do here:

1. "Preview & Send" button should take you to a read-only preview of your message, from which you can either go back to edit, or continue forward to send (which we can just disable for now, no need even for a "success" page) – the aim of this "Preview" step is to give people a chance to read what they’ve written, with fresh eyes, sort of, and maybe re-think any unfriendly content with a cooler head!
2. I’ve added a link to the form, from a contract page. I wonder whether it’s possible to "carry through" the information about that contract, _onto_ this form page, so people can see the contract details (maybe name of contract, time left, cost – all the stuff you’d see in the "summary" on the homepage / search results page) while they’re writing?
3. Would be nice if the Email form controls were `type="email"` instead of `type="text"` so that touchscreen devices will display the right keyboard.

alexander-griffen committed:
Alter form to make email fields produce the correct keyboard on phones.

struan commented on procurement/management/commands/import_climate_representatives.py:
```diff
@@ -0,0 +1,100 @@
+                        council=council,
+                        job_title=job_title,
+                        party_affiliation=party_affiliation,
+                        slug=ClimateCouncillor.slugify_name(title + ' ' + first_name + ' ' + surname),
```
What if there's more than one person with the same name? I suspect this might need something else for looking things up.

alexander-griffen replied:
They have been changed to be a string of the primary key.

struan commented on procurement/forms.py:
```diff
@@ -29,14 +29,56 @@ def clean_pc(self):
+    pc = forms.CharField(widget=forms.TextInput(attrs={"class": "form-control"}))
+
+
+class ContactRepresentativeForm(PostcodeForm):
```
yay for shared code!

struan commented on procurement/forms.py:
```diff
@@ -29,14 +29,56 @@ def clean_pc(self):
+    # Use CharField + TextInput for the email boxes so that the
+    # validator raises errors rather than using the browser validation,
+    # for consistency.
+    email = forms.CharField(widget=forms.TextInput(), validators=[EmailValidator()])
```
and yay for super helpful comments :)

struan commented on procurement/views.py:
```diff
@@ -144,3 +144,41 @@ def get_queryset(self):
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        council = get_object_or_404(Council.objects.all(), slug=self.kwargs["council"])      
```
You can use the model class, e.g.`Council`, as the first arg to `get_objects_or_404`

struan commented on procurement/views.py:
```diff
@@ -144,3 +144,41 @@ def get_queryset(self):
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        get_object_or_404(Council.objects.all(), slug=self.kwargs["council"])  
+        context["representative"] = get_object_or_404(ClimateRepresentative.objects.all(), slug=self.kwargs["representative"])
```
This should probably check that the person is part of the council too.

struan commented on procurement/tests/data/climate_representative_import/import_climate_councillors_bad_data.csv:
```diff
@@ -0,0 +1,4 @@
+title,firstName,surname,type,represents,ward,council,councilInternalName,region,county,gssNumber,partyAffiliation,email,code,councilGssNumber,countyGssNumber,genericPositionLocalList,secondEmail,local-authority-code,official-name,local-authority-type-name
+Cllr,Eamon,McNeill,Councillor,"Portadown in Armagh, Banbridge and Craigavon",Portadown,"Armagh, Banbridge and Craigavon","Armagh City, Banbridge and Craigavon Borough Council",Northern Ireland,Northern Ireland,N10000207,SDLP,eamon.mcneill@armaghbanbridgecraigavon.gov.uk,10056588,N09000002,N92000002,Chairman Climate Change (Environmental Services Committee),,ABC,"Armagh City, Banbridge and Craigav ...
+,,,,,,,,,,,,,,,,,,NECA,North East Combined Authority,Combined authority
```
I realise this is going to be a bit of a pain but we should probably use fake names/emails here.

struan commented on procurement/tests/test_import_climate_representatives.py:
```diff
@@ -0,0 +1,150 @@
+    def setUp(self):
+        call_command(
+            "import_councils",
+        )
```
This should ideally use a fixture to add the required councils rather than running the import command.

struan reviewed (changes requested):
This looks basically good, few small things and one test data thing.

struan commented on procurement/models.py:
```diff
@@ -179,3 +179,26 @@ def contract_time_remaining_desc(self):
             return "{} days left".format(timedelta.days)
+
+
+class ClimateRepresentative(models.Model):
```
This might want a
```
class Meta:
    abstract = True
```
in it otherwise you end up with a table for this plus tables for the inheriting classes which I'm not sure is what you want? There's some more details on this here: https://docs.djangoproject.com/en/4.1/topics/db/models/#model-inheritance

alexander-griffen replied:
Thanks, I missed this!

alexander-griffen committed:
Carry the contract information through to the representative page.

alexander-griffen committed:
Implement preview page for representative contact form.

struan reviewed (approved):
Looks good.

zarino reviewed (approved):
Looks good to me. Definitely needs a bit of UI design love – I will sort that out.

@alexander-griffen happy for me to keep this branch/PR open for a few days while I make those changes? You can unassign yourself as I think your work here is done 🙂

alexander-griffen commented:
> @alexander-griffen happy for me to keep this branch/PR open for a few days while I make those changes? You can unassign yourself as I think your work here is done 🙂

Sounds good, thank you!

zarino committed:
Populate <title> tag with page_title context variable

Also includes a ClimateRepresentative.full_name property, to DRY up the
various places we were constructing representatives’ full names.
